apps/astarte_pairing/lib/astarte_pairing/fdo/voucher_manager.ex

```diff
@@ -0,0 +1,52 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2025 SECO Mind Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.Pairing.FDO.VoucherManager do
+  alias Astarte.Pairing.Queries
+  require Logger
+
+  def save_voucher(realm_name, voucher_blob, pkey_blob) do
+    with {:ok, inserted_voucher} <-
+           Queries.create_ownership_voucher(realm_name, voucher_blob, pkey_blob),
+         device_guuid <- extract_device_guuid(voucher_blob),
+         {:ok, _} <-
+           Queries.update_device_registration_voucher(
+             realm_name,
+             device_guuid,
+             inserted_voucher.id
+           ) do
+      :ok
+    end
+  end
+
+  defp extract_device_guuid(voucher_blob) do
+    with ov_data <-
+           voucher_blob
+           |> String.replace("-----BEGIN OWNERSHIP VOUCHER-----", "")
+           |> String.replace("-----END OWNERSHIP VOUCHER-----", "")
+           |> String.replace(~r/\s/, ""),
+         {:ok, cbor_data} <- Base.decode64(ov_data),
+         {:ok, decoded_payload, _rest} <- cbor_data |> CBOR.decode(),
+         {:ok, decoded_header, _rest} <- Enum.at(decoded_payload, 1).value |> CBOR.decode(),
+         device_guid <- Enum.at(decoded_header, 1).value |> UUID.binary_to_string!() do
+      device_guid
```

Comment:
ok so this is a lot to unpack

for starters, this is out of scope for the with as you're matching the whole thing (like `ov_data`)

```suggestion
      ...   do
      decoded_header
      |> Enum.at(1)
      |> Map.fetch!(:value) 
      |> UUID.binary_to_string!() 
```

then, is there a reason why we're grabbing the second element?

lastly, for clarity, prefer having an helper function which describes the logic a little


```suggestion
      ...   do
      device_id = device_id_from_header(decoded_header)
      UUID.binary_to_string!(device_id)
```